diff -ur mutt-1.13.3.orig/globals.h mutt-1.13.3/globals.h
--- mutt-1.13.3.orig/globals.h	2019-11-30 15:59:38.000000000 -0600
+++ mutt-1.13.3/globals.h	2020-02-01 13:49:44.023625247 -0600
@@ -129,6 +129,7 @@
 WHERE char *QueryCmd;
 WHERE char *QueryFormat;
 WHERE char *Realname;
+WHERE char *ReplyFilter;
 WHERE short SearchContext;
 WHERE char *SendCharset;
 WHERE char *SendMultipartAltFilter;
diff -ur mutt-1.13.3.orig/init.h mutt-1.13.3/init.h
--- mutt-1.13.3.orig/init.h	2020-01-12 14:17:28.000000000 -0600
+++ mutt-1.13.3/init.h	2020-02-01 13:52:38.630120333 -0600
@@ -2914,6 +2914,13 @@
   ** .pp
   ** Also see $$wrap.
   */
+  { "reply_filter",	DT_PATH, R_NONE, {.p=&ReplyFilter}, {.p=0} },
+  /*
+  ** .pp
+  ** When set, specifies a command used to filter quoted replies.  When a quoted
+  ** message is about to be included in a reply it is passed as standard input to
+  ** $$reply_filter, and the text to include is read from the standard output.
+  */
   { "reply_regexp",	DT_RX,	 R_INDEX|R_RESORT, {.p=&ReplyRegexp}, {.p="^(re([\\[0-9\\]+])*|aw):[ \t]*"} },
   /*
   ** .pp
diff -ur mutt-1.13.3.orig/send.c mutt-1.13.3/send.c
--- mutt-1.13.3.orig/send.c	2019-12-03 20:21:46.000000000 -0600
+++ mutt-1.13.3/send.c	2020-02-01 13:49:44.023625247 -0600
@@ -885,6 +885,8 @@
   int i;
   HEADER *h;
   BODY *tmp;
+  FILE *filterfp = NULL;
+  pid_t filterpid = -1;
 
   if (flags & SENDREPLY)
   {
@@ -894,6 +896,18 @@
     if (i == MUTT_YES)
     {
       mutt_message _("Including quoted message...");
+      if (ReplyFilter && *ReplyFilter) 
+      {
+	filterfp = tempfp;
+	tempfp = NULL;
+	filterpid = mutt_create_filter_fd (ReplyFilter, &tempfp, NULL, NULL,
+					   -1, fileno(filterfp), -1);
+	if (filterpid < 0)
+	{
+	  mutt_error (_("Could not create reply filter"));
+	  return (-1);
+	}
+      }
       if (!cur)
       {
 	for (i = 0; i < ctx->vcount; i++)
@@ -913,6 +927,13 @@
       else
 	include_reply (ctx, cur, tempfp);
 
+      if (filterfp != NULL)
+      {
+	safe_fclose(&tempfp);
+	if (mutt_wait_filter (filterpid) != 0)
+	  mutt_any_key_to_continue (NULL);
+	tempfp = filterfp;
+      }
     }
   }
   else if (flags & SENDFORWARD)
