diff -Nru xscreensaver-6.02.orig/driver/XScreenSaver.ad.in xscreensaver-6.02/driver/XScreenSaver.ad.in
--- xscreensaver-6.02.orig/driver/XScreenSaver.ad.in	2021-10-11 13:49:37.173116852 -0500
+++ xscreensaver-6.02/driver/XScreenSaver.ad.in	2022-02-17 08:36:58.728285873 -0600
@@ -327,6 +327,7 @@
 				maze -root				    \n\
 @GL_KLUDGE@ GL: 				superquadrics -root			    \n\
 				attraction -root			    \n\
+				adstarry -root				    \n\
 				blitspin -root				    \n\
 				greynetic -root				    \n\
 				helix -root				    \n\
diff -Nru xscreensaver-6.02.orig/hacks/Makefile.in xscreensaver-6.02/hacks/Makefile.in
--- xscreensaver-6.02.orig/hacks/Makefile.in	2021-10-11 20:04:29.723043835 -0500
+++ xscreensaver-6.02/hacks/Makefile.in	2022-02-17 08:38:32.397963019 -0600
@@ -92,6 +92,7 @@
 		  $(UTILS_BIN)/font-retry.o
 
 SRCS		= xscreensaver-getimage.c \
+		  adstarry.c \
 		  attraction.c blitspin.c bouboule.c braid.c bubbles.c \
 		  bubbles-default.c decayscreen.c deco.c drift.c flag.c \
 		  flame.c forest.c vines.c galaxy.c grav.c greynetic.c \
@@ -132,6 +133,7 @@
 # but that are not shipped with xscreensaver itself.
 #
 OBJS		= attraction.o blitspin.o bouboule.o braid.o bubbles.o \
+		  adstarry.o \
 		  bubbles-default.o decayscreen.o deco.o drift.o flag.o \
 		  flame.o forest.o vines.o galaxy.o grav.o greynetic.o \
 		  halo.o helix.o hopalong.o hypercube.o ifs.o imsmap.o \
@@ -194,6 +196,8 @@
 		  lightning lisa lissie lmorph rotor sphere spiral t3d vines \
 		  whirlygig worm xsublim juggle
 
+EXES           += adstarry
+
 HACK_OBJS_1	= fps.o $(UTILS_BIN)/resources.o $(UTILS_BIN)/visual.o \
 		  $(UTILS_BIN)/usleep.o $(UTILS_BIN)/yarandom.o \
 		  $(UTILS_BIN)/utf8wc.o $(UTILS_BIN)/font-retry.o \
@@ -578,6 +582,9 @@
 deco:		deco.o		$(HACK_OBJS) $(COL)
 	$(CC_HACK) -o $@ $@.o	$(HACK_OBJS) $(COL) $(HACK_LIBS)
 
+adstarry:	adstarry.o	$(HACK_OBJS) $(PNG)
+	$(CC_HACK) -o $@ $@.o	$(HACK_OBJS) $(PNG) $(PNG_LIBS)
+
 flame:		flame.o		$(HACK_OBJS) $(COL)
 	$(CC_HACK) -o $@ $@.o	$(HACK_OBJS) $(COL) $(HACK_LIBS)
 
diff -Nru xscreensaver-6.02.orig/hacks/adstarry.c xscreensaver-6.02/hacks/adstarry.c
--- xscreensaver-6.02.orig/hacks/adstarry.c	1969-12-31 18:00:00.000000000 -0600
+++ xscreensaver-6.02/hacks/adstarry.c	2022-02-17 08:36:58.731619166 -0600
@@ -0,0 +1,331 @@
+/* Starry Night by Craig Dickson, Copyright (c) 1990-1994 Berkeley Systems Inc.
+ *
+ * Permission to use, copy, modify, distribute, and sell this software and its
+ * documentation for any purpose is so totally not granted.
+ */
+
+#include "screenhack.h"
+
+#include <stdlib.h>
+#include <sys/time.h>
+
+#include "ximage-loader.h"
+#include "images/gen/meteor_l_png.h"
+#include "images/gen/meteor_r_png.h"
+#include "images/gen/flasher_png.h"
+
+
+#define BOUND(a, x, b) MIN(MAX((a),(x)),(b))
+
+static inline int
+MOD(int a, int b)
+{
+	return b ? (a % b) : 0;
+}
+
+static inline int
+MIN(int a, int b)
+{
+	return (a < b) ? a : b;
+}
+
+static inline int
+MAX(int a, int b)
+{
+	return (a > b) ? a : b;
+}
+
+
+struct state {
+	XWindowAttributes wa;
+	GC gc;
+
+	int numBuildings;
+	int bldgHeight;
+	Bool flasher;
+
+	unsigned long lastFlash;
+
+	Pixmap meteorPM[2];
+	int meteorW, meteorH;
+	int meteorX, meteorY;
+	Pixmap flasherPM;
+	int flasherW, flasherH;
+	int flasherX, flasherY;
+
+	int *bldgTop, *bldgLeft, *bldgRight, *bldgPat;
+	/*
+	XColor colors[255];
+	int ncolors;
+	int max_depth;
+	int min_height;
+	int min_width;
+	int line_width;
+	int old_line_width;
+	Bool goldenRatio;
+	Bool mondrian;
+	Bool smoothColors;
+
+	int delay;
+	GC fgc, bgc;
+	int current_color;
+	*/
+};
+
+
+static unsigned char patterns[][8] = {
+	{0x00, 0x00, 0x00, 0x66, 0x66, 0x00, 0x00, 0x00},
+	{0x00, 0x00, 0x44, 0x44, 0x00, 0x00, 0x44, 0x44},
+	{0x88, 0x88, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
+	{0xc6, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xc6},
+	{0x00, 0x00, 0x54, 0x54, 0x54, 0x00, 0x00, 0x00},
+	{0x00, 0x00, 0x00, 0x48, 0x48, 0x48, 0x48, 0x00},
+	{0x00, 0x00, 0x00, 0x54, 0x00, 0x00, 0x00, 0x00},
+	{0x01, 0x01, 0x00, 0x00, 0x01, 0x01, 0x00, 0x00},
+};
+
+
+static void
+load_pixmaps(Display *dpy, Window win, struct state *s)
+{
+	s->meteorPM[0] = image_data_to_pixmap(dpy, win, meteor_l_png,
+			sizeof(meteor_l_png), &s->meteorW, &s->meteorH, 0);
+	s->meteorPM[1] = image_data_to_pixmap(dpy, win, meteor_r_png,
+			sizeof(meteor_r_png), &s->meteorW, &s->meteorH, 0);
+	s->flasherPM = image_data_to_pixmap(dpy, win, flasher_png,
+			sizeof(flasher_png), &s->flasherW, &s->flasherH, 0);
+}
+
+static int
+weighted_random(int max)
+{
+	return max ? MIN(random() % max, random() % max) : 0;
+}
+
+static int
+find_bldg(struct state *s, int x, int y)
+{
+	int b;
+
+	for (b = 0; b < s->numBuildings; b++)
+		if (x >= s->bldgLeft[b] && x <= s->bldgRight[b] &&
+				y >= s->bldgTop[b])
+			return b;
+	return -1;
+}
+
+static void
+draw_star(Display *dpy, Window win, struct state *s)
+{
+	unsigned long color;
+	int x, y;
+
+	x = MOD(random(), s->wa.width);
+	y = weighted_random(s->wa.height);
+
+	if (find_bldg(s, x, y) != -1)
+		return;
+	color  = (random() % 256) << 16;
+	color |= (random() % 256) << 8;
+	color |= (random() % 256);
+	XSetForeground(dpy, s->gc, color);
+	XDrawPoint(dpy, win, s->gc, x, y);
+}
+
+static void
+draw_bldg_px(Display *dpy, Window win, struct state *s,
+		int x, int y, int bldg)
+{
+	unsigned char pat_byte;
+	int brightness, subx, suby;
+	unsigned long color;
+
+	if (s->bldgLeft[bldg] > x || s->bldgRight[bldg] < x ||
+			s->bldgTop[bldg] > y)
+		return;
+
+	brightness = 255 - 192 / s->numBuildings * bldg;
+	subx = (x + s->bldgLeft[bldg]) & 7;
+	suby = (y + s->bldgTop[bldg]) & 7;
+	pat_byte = patterns[s->bldgPat[bldg]][suby];
+	if (!((pat_byte >> subx) & 1))
+		return;
+	color = brightness << 16 | brightness << 8;
+	XSetForeground(dpy, s->gc, color);
+	XDrawPoint(dpy, win, s->gc, x, y);
+}
+
+static void
+draw_bldg_block(Display *dpy, Window win, struct state *s)
+{
+	int x, y, dx, dy;
+	int bldg;
+
+	x = MOD(random(), s->wa.width);
+	y = MOD(random(), s->wa.height);
+	bldg = find_bldg(s, x, y);
+	if (bldg == -1)
+		return;
+	for (dy = -1; dy < 2; dy++)
+		for (dx = -1; dx < 2; dx++)
+			draw_bldg_px(dpy, win, s, x + dx, y + dy, bldg);
+}
+
+static void
+update_flasher(Display *dpy, Window win, struct state *s)
+{
+	struct timeval t;
+	long ms;
+
+	if (!s->flasher)
+		return;
+
+	gettimeofday(&t, NULL);
+	ms = t.tv_sec * 1000 + t.tv_usec / 1000;
+	if (ms - s->lastFlash <= 750)
+		return;
+
+	s->lastFlash = ms;
+	XCopyArea(dpy, s->flasherPM, win, s->gc, 0, 0, s->flasherW, s->flasherH,
+			s->flasherX, s->flasherY);
+}
+
+static void
+update_meteor(Display *bpy, Window win, struct state *s)
+{
+}
+
+static void
+init_bldgs(struct state *s)
+{
+	int i, maxHeight;
+	int width, center;
+
+	s->bldgTop = realloc(s->bldgTop, s->numBuildings * sizeof *s->bldgTop);
+	s->bldgLeft = realloc(s->bldgLeft, s->numBuildings * sizeof *s->bldgLeft);
+	s->bldgRight = realloc(s->bldgRight, s->numBuildings * sizeof *s->bldgRight);
+	s->bldgPat = realloc(s->bldgPat, s->numBuildings * sizeof *s->bldgPat);
+
+	maxHeight = MAX(20, s->wa.height * s->bldgHeight / 100);
+
+	s->flasherX = 0;
+	s->flasherY = s->wa.height;
+
+	for (i = 0; i < s->numBuildings; i++) {
+		s->bldgPat[i] = random() & 7;
+		s->bldgTop[i] = s->wa.height - random() % maxHeight;
+
+		center = MOD(random(), s->wa.width);
+
+		if (s->bldgTop[i] - 4 < s->flasherY) {
+			s->flasherY = s->bldgTop[i] - 4;
+			s->flasherX = center - 2;
+		}
+
+		width = weighted_random(s->bldgTop[i] / 8);
+		width += s->bldgTop[i] / 16 + 8;
+		width = MAX(16, width);
+		s->bldgLeft[i] = center - width / 2;
+		s->bldgRight[i] = s->bldgLeft[i] + width;
+	}
+}
+
+static void *
+starry_init(Display *dpy, Window win)
+{
+	struct state *s = calloc(1, sizeof *s);
+	XGCValues gcv;
+
+	/* get configurables */
+	s->numBuildings = get_integer_resource(dpy, "numBuildings", "Integer");
+	s->bldgHeight = get_integer_resource(dpy, "bldgHeight", "Integer");
+	s->flasher = get_boolean_resource(dpy, "flasher", "Boolean");
+
+	XGetWindowAttributes(dpy, win, &s->wa);
+
+	/* set up buildings */
+	init_bldgs(s);
+
+	/* load meteor images */
+	load_pixmaps(dpy, win, s);
+
+	/* init colors?? */
+
+	/* set up GC */
+	gcv.function = GXxor;
+	gcv.foreground = WhitePixel(dpy, DefaultScreen(dpy));
+	gcv.background = BlackPixel(dpy, DefaultScreen(dpy));
+	s->gc = XCreateGC(dpy, win,
+			GCFunction | GCForeground | GCBackground, &gcv);
+
+	return s;
+}
+
+static unsigned long
+starry_draw(Display *dpy, Window win, void *state)
+{
+	struct state *s = (struct state *) state;
+	int i;
+	int x, y;
+
+	draw_star(dpy, win, s);
+	XSetForeground(dpy, s->gc, 0);
+	for (i = 0; i < 50; i++) {
+		x = MOD(random(), s->wa.width);
+		y = weighted_random(s->wa.height);
+		XDrawPoint(dpy, win, s->gc, x, y);
+	}
+	for (i = 0; i < 30; i++) {
+		x = MOD(random(), s->wa.width);
+		y = MOD(random(), s->wa.height);
+		XDrawPoint(dpy, win, s->gc, x, y);
+	}
+	for (i = 0; i < 15; i++)
+		draw_bldg_block(dpy, win, s);
+	update_flasher(dpy, win, s);
+	update_meteor(dpy, win, s);
+	return 10000;
+}
+
+static void
+starry_reshape(Display *dpy, Window win, void *state,
+		unsigned int w, unsigned int h)
+{
+	struct state *s = (struct state *) state;
+	s->wa.width = w;
+	s->wa.height = h;
+}
+
+static Bool
+starry_event(Display *dpy, Window win, void *state, XEvent *event)
+{
+	return False;
+}
+
+static void
+starry_free(Display *dpy, Window win, void *state)
+{
+	struct state *s = state;
+	free(s->bldgPat);
+	free(s->bldgRight);
+	free(s->bldgLeft);
+	free(s->bldgTop);
+	free(s);
+}
+
+static const char *starry_defaults [] = {
+	"*numBuildings:		27",
+	"*bldgHeight:		34",
+	"*flasher:		True",
+	0
+};
+
+static XrmOptionDescRec starry_options [] = {
+	{ "-num-buildings", ".numBuildings", XrmoptionSepArg, 0 },
+	{ "-bldg-height", ".bldgHeight", XrmoptionSepArg, 0 },
+	{ "-flasher", ".flasher", XrmoptionNoArg, "True" },
+	{ "-no-flasher", ".flasher", XrmoptionNoArg, "False" },
+	{ 0, 0, 0, 0 }
+};
+
+XSCREENSAVER_MODULE("StarryNight", starry)
diff -Nru xscreensaver-6.02.orig/hacks/config/adstarry.xml xscreensaver-6.02/hacks/config/adstarry.xml
--- xscreensaver-6.02.orig/hacks/config/adstarry.xml	1969-12-31 18:00:00.000000000 -0600
+++ xscreensaver-6.02/hacks/config/adstarry.xml	2022-02-17 08:36:58.731619166 -0600
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+
+<screensaver name="adstarry" _label="StarryNight">
+
+  <command arg="-root"/>
+
+  <number id="numBuildings" type="slider" arg="-num-buildings %"
+          _label="Buildings" _low-label="0" _high-label="100"
+          low="0" high="100" default="27"/>
+
+  <number id="bldgHeight" type="slider" arg="-bldg-height %"
+            _label="Bldg Height" _low-label="5%" _high-label="95%"
+            low="5" high="95" default="34"/>
+
+  <boolean id="flasher" _label="Flasher" arg-unset="-no-flasher"/>
+
+  <_description>
+STARRY NIGHT
+
+Original concept by James J. Eastman
+
+The number of buildings can be varied from 0 to 100.  The maximum height
+of the buildings can be varied from 5% to 95% of the monitor height.  The
+'flasher' is a flashing red light which, if enabled, will appear atop the
+tallest building.
+  </_description>
+</screensaver>
